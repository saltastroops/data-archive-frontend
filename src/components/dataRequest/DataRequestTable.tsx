import moment from "moment";
import * as React from "react";
import styled from "styled-components";
import { IDataRequest, IDataRequestPart } from "./DataRequestsForm";
import DataRequestTableRow from "./DataRequestTableRow";

/**
 * Properties of the data request table component.
 *
 * Properties:
 * -----------
 * dataRequest:
 *    Data request whose content is displayed.
 */
interface IDataRequestTableProps {
  dataRequest: IDataRequest;
}

const Table = styled.table.attrs({
  className: "table is-striped is-narrowed is-hoverable is-fullwidth"
})`
  && {
  }
`;

/**
 * The string to display for the creation date.
 *
 * If the date lies within the past 24 hours, the difference from now (such as
 * "5 seconds ago" or "17 hours ago") is returned. Otherwise the time is
 * returned as date and time (such as "on 21 Feb 2019 at 2:16").
 *
 * Parameters
 * ----------
 * madeAt:
 *    The creation date as a string that can be parsed by MomentJS.
 */
const displayedTime = (madeAt: string) => {
  const t = moment(madeAt);
  if (Date.now() - t.valueOf() < 24 * 3600 * 1000) {
    return t.fromNow();
  } else {
    return `on ${t.format("D MMM YYYY")} at ${t.format("k:mm")}`;
  }
};

/**
 * A table displaying the content of a data request.
 */
class DataRequestTable extends React.Component<IDataRequestTableProps> {
  render() {
    const { parts, madeAt, uri } = this.props.dataRequest;

    const mayDownloadAll = parts.every(part => part.status === "SUCCESSFUL");

    const reRequestAll =
      !mayDownloadAll && !parts.some(part => part.status === "PENDING");

    return (
      <Table>
        <thead>
          <tr>
            <th colSpan={3}>
              <p style={{ display: "inline-block" }}>
                Requested {displayedTime(madeAt)}
              </p>
              <p
                style={{
                  display: "inline-block",
                  position: "absolute",
                  right: 0
                }}
              >
                {mayDownloadAll && (
                  <a
                    className="button download-all is-small is-success is-rounded"
                    href={uri}
                  >
                    Download all
                  </a>
                )}
                {reRequestAll && (
                  <button className="button re-request-all is-small is-danger is-rounded">
                    Re-request all
                  </button>
                )}
              </p>
            </th>
          </tr>
          <tr>
            <th>Observation</th>
            <th>Status</th>
            <th>Files</th>
          </tr>
        </thead>
        <tbody>
          {parts.map((part: IDataRequestPart) => {
            return <DataRequestTableRow key={part.id} dataRequestPart={part} />;
          })}
        </tbody>
      </Table>
    );
  }
}

export default DataRequestTable;
