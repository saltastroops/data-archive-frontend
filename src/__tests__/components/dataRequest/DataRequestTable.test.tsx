import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import {
  DataRequestStatus,
  IDataRequest
} from "../../../components/dataRequest/DataRequestsForm";
import DataRequestTable from "../../../components/dataRequest/DataRequestTable";

const dummyDataRequestsWrapper = (...statusValues: DataRequestStatus[]) => {
  const parts = statusValues.map((status, index) => ({
    dataFiles: [
      {
        id: `file-${index}a`,
        name: `filename ${index}a`,
        observation: {
          id: `obs-${index}a`,
          name: `Observation ${index}a`
        }
      },
      {
        id: `file-${index}b`,
        name: `filename ${index}b`,
        observation: {
          id: `obs-${index}b`,
          name: `Observation ${index}b`
        }
      }
    ],
    id: `part-${index}`,
    status,
    uri: `http://demo/data-request/part/part-${index}`
  }));

  const dataReqest = {
    id: "request-1",
    madeAt: "2018-12-01 20:00",
    parts,
    uri: `http://demo/data-request/request-1`
  } as IDataRequest;

  return mount(<DataRequestTable dataRequest={dataReqest} />);
};

describe("DataRequestTable", () => {
  it("should render correctly", () => {
    expect(
      toJson(dummyDataRequestsWrapper("SUCCESSFUL", "FAILED", "EXPIRED"))
    ).toMatchSnapshot();
  });

  it("includes a link for downloading all files if all request parts are marked successful", async () => {
    // all parts are marked successful

    let wrapper = dummyDataRequestsWrapper("SUCCESSFUL");
    expect(wrapper.find(".download-all").length).toBe(1);

    wrapper = dummyDataRequestsWrapper("SUCCESSFUL", "SUCCESSFUL");
    expect(wrapper.find(".download-all").length).toBe(1);

    wrapper = dummyDataRequestsWrapper(
      "SUCCESSFUL",
      "SUCCESSFUL",
      "SUCCESSFUL"
    );
    expect(wrapper.find(".download-all").length).toBe(1);
  });

  it("does not include a link for downloading all files if at least one request part is not marked successful", async () => {
    let wrapper = dummyDataRequestsWrapper("PENDING");
    expect(wrapper.find(".download-all").length).toBe(0);

    wrapper = dummyDataRequestsWrapper("FAILED");
    expect(wrapper.find(".download-all").length).toBe(0);

    wrapper = dummyDataRequestsWrapper("EXPIRED");
    expect(wrapper.find(".download-all").length).toBe(0);

    wrapper = dummyDataRequestsWrapper("EXPIRED", "PENDING");
    expect(wrapper.find(".download-all").length).toBe(0);

    wrapper = dummyDataRequestsWrapper("SUCCESSFUL", "PENDING", "SUCCESSFUL");
    expect(wrapper.find(".download-all").length).toBe(0);
  });

  it("includes a link for re-requesting all files if not all parts are marked successful and none is pending", () => {
    let wrapper = dummyDataRequestsWrapper("EXPIRED");
    expect(wrapper.find(".re-request-all").length).toBe(1);

    wrapper = dummyDataRequestsWrapper("FAILED");
    expect(wrapper.find(".re-request-all").length).toBe(1);

    wrapper = dummyDataRequestsWrapper("SUCCESSFUL", "FAILED");
    expect(wrapper.find(".re-request-all").length).toBe(1);

    wrapper = dummyDataRequestsWrapper("EXPIRED", "SUCCESSFUL", "EXPIRED");
    expect(wrapper.find(".re-request-all").length).toBe(1);
  });

  it("does not include a link for re-requesting all files if all parts are marked successful or there is a pending one", () => {
    let wrapper = dummyDataRequestsWrapper("SUCCESSFUL");
    expect(wrapper.find(".re-request-all").length).toBe(0);

    wrapper = dummyDataRequestsWrapper("SUCCESSFUL", "SUCCESSFUL");
    expect(wrapper.find(".re-request-all").length).toBe(0);

    wrapper = dummyDataRequestsWrapper(
      "SUCCESSFUL",
      "SUCCESSFUL",
      "SUCCESSFUL"
    );
    expect(wrapper.find(".re-request-all").length).toBe(0);

    wrapper = dummyDataRequestsWrapper("PENDING");
    expect(wrapper.find(".re-request-all").length).toBe(0);

    wrapper = dummyDataRequestsWrapper("FAILED", "EXPIRED", "PENDING");
    expect(wrapper.find(".re-request-all").length).toBe(0);
  });
});
