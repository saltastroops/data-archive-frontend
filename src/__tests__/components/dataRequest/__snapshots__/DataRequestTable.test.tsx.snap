// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataRequestTable should render correctly 1`] = `
<DataRequestTable
  dataRequest={
    Object {
      "id": "request-1",
      "madeAt": "2018-12-01 20:00",
      "parts": Array [
        Object {
          "dataFiles": Array [
            Object {
              "id": "file-0a",
              "name": "filename 0a",
              "observation": Object {
                "id": "obs-0a",
                "name": "Observation 0a",
              },
            },
            Object {
              "id": "file-0b",
              "name": "filename 0b",
              "observation": Object {
                "id": "obs-0b",
                "name": "Observation 0b",
              },
            },
          ],
          "id": "part-0",
          "status": "SUCCESSFUL",
          "uri": "http://demo/data-request/part/part-0",
        },
        Object {
          "dataFiles": Array [
            Object {
              "id": "file-1a",
              "name": "filename 1a",
              "observation": Object {
                "id": "obs-1a",
                "name": "Observation 1a",
              },
            },
            Object {
              "id": "file-1b",
              "name": "filename 1b",
              "observation": Object {
                "id": "obs-1b",
                "name": "Observation 1b",
              },
            },
          ],
          "id": "part-1",
          "status": "FAILED",
          "uri": "http://demo/data-request/part/part-1",
        },
        Object {
          "dataFiles": Array [
            Object {
              "id": "file-2a",
              "name": "filename 2a",
              "observation": Object {
                "id": "obs-2a",
                "name": "Observation 2a",
              },
            },
            Object {
              "id": "file-2b",
              "name": "filename 2b",
              "observation": Object {
                "id": "obs-2b",
                "name": "Observation 2b",
              },
            },
          ],
          "id": "part-2",
          "status": "EXPIRED",
          "uri": "http://demo/data-request/part/part-2",
        },
      ],
      "uri": "http://demo/data-request/request-1",
    }
  }
>
  <styled.table>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [
            Object {
              "className": "table is-striped is-narrowed is-hoverable is-fullwidth",
            },
          ],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": true,
            "lastClassName": "dPKlrP",
            "rules": Array [
              "
  && {
  }
",
            ],
          },
          "displayName": "styled.table",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "table",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <table
        className="table is-striped is-narrowed is-hoverable is-fullwidth "
      >
        <thead>
          <tr>
            <th
              colSpan={3}
            >
              <p
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                Requested 
                on 1 Dec 2018 at 20:00
              </p>
              <p
                style={
                  Object {
                    "display": "inline-block",
                    "position": "absolute",
                    "right": 0,
                  }
                }
              >
                <button
                  className="button re-request-all is-small is-danger is-rounded"
                >
                  Re-request all
                </button>
              </p>
            </th>
          </tr>
          <tr>
            <th>
              Observation
            </th>
            <th>
              Status
            </th>
            <th>
              Files
            </th>
          </tr>
        </thead>
        <tbody>
          <DataRequestTableRow
            dataRequestPart={
              Object {
                "dataFiles": Array [
                  Object {
                    "id": "file-0a",
                    "name": "filename 0a",
                    "observation": Object {
                      "id": "obs-0a",
                      "name": "Observation 0a",
                    },
                  },
                  Object {
                    "id": "file-0b",
                    "name": "filename 0b",
                    "observation": Object {
                      "id": "obs-0b",
                      "name": "Observation 0b",
                    },
                  },
                ],
                "id": "part-0",
                "status": "SUCCESSFUL",
                "uri": "http://demo/data-request/part/part-0",
              }
            }
            key="part-0"
          >
            <tr>
              <td>
                Observation 0a
              </td>
              <td>
                <p>
                  Available
                </p>
                <p>
                  <a
                    className="button download is-small is-success is-rounded"
                    href="http://demo/data-request/part/part-0"
                  >
                    Download
                  </a>
                </p>
              </td>
              <td>
                <ul>
                  <li
                    key="file-0a"
                  >
                    filename 0a
                  </li>
                  <li
                    key="file-0b"
                  >
                    filename 0b
                  </li>
                </ul>
              </td>
            </tr>
          </DataRequestTableRow>
          <DataRequestTableRow
            dataRequestPart={
              Object {
                "dataFiles": Array [
                  Object {
                    "id": "file-1a",
                    "name": "filename 1a",
                    "observation": Object {
                      "id": "obs-1a",
                      "name": "Observation 1a",
                    },
                  },
                  Object {
                    "id": "file-1b",
                    "name": "filename 1b",
                    "observation": Object {
                      "id": "obs-1b",
                      "name": "Observation 1b",
                    },
                  },
                ],
                "id": "part-1",
                "status": "FAILED",
                "uri": "http://demo/data-request/part/part-1",
              }
            }
            key="part-1"
          >
            <tr>
              <td>
                Observation 1a
              </td>
              <td>
                <p>
                  Failed
                </p>
                <p>
                  <button
                    className="button re-request is-small is-danger is-rounded"
                  >
                    Re-request
                  </button>
                </p>
              </td>
              <td>
                <ul>
                  <li
                    key="file-1a"
                  >
                    filename 1a
                  </li>
                  <li
                    key="file-1b"
                  >
                    filename 1b
                  </li>
                </ul>
              </td>
            </tr>
          </DataRequestTableRow>
          <DataRequestTableRow
            dataRequestPart={
              Object {
                "dataFiles": Array [
                  Object {
                    "id": "file-2a",
                    "name": "filename 2a",
                    "observation": Object {
                      "id": "obs-2a",
                      "name": "Observation 2a",
                    },
                  },
                  Object {
                    "id": "file-2b",
                    "name": "filename 2b",
                    "observation": Object {
                      "id": "obs-2b",
                      "name": "Observation 2b",
                    },
                  },
                ],
                "id": "part-2",
                "status": "EXPIRED",
                "uri": "http://demo/data-request/part/part-2",
              }
            }
            key="part-2"
          >
            <tr>
              <td>
                Observation 2a
              </td>
              <td>
                <p>
                  Expired
                </p>
                <p>
                  <button
                    className="button re-request is-small is-danger is-rounded"
                  >
                    Re-request
                  </button>
                </p>
              </td>
              <td>
                <ul>
                  <li
                    key="file-2a"
                  >
                    filename 2a
                  </li>
                  <li
                    key="file-2b"
                  >
                    filename 2b
                  </li>
                </ul>
              </td>
            </tr>
          </DataRequestTableRow>
        </tbody>
      </table>
    </StyledComponent>
  </styled.table>
</DataRequestTable>
`;
